File: loupedeck-plugin/ApricadabraPlugin/src/CoreConnection.cs
using System;
using System.Diagnostics;
using System.IO;
using System.IO.Pipes;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Threading;
using System.Threading.Tasks;

namespace Loupedeck.ApricadabraPlugin
{
    public class CoreConnection : IDisposable
    {
        private const string PipeName = "apricadabra";
        private const string CoreExeName = "apricadabra-core.exe";
        private const int ProtocolVersion = 2;
        private const int UdpCommandPort = 19871;
        private const int UdpBroadcastPort = 19872;

        private NamedPipeClientStream _pipe;
        private StreamReader _reader;
        private StreamWriter _writer;
        private UdpClient _udpSender;
        private CancellationTokenSource _cts;
        private Task _pipeReadTask;
        private Task _udpListenTask;
        private bool _connected;
        private DateTime _coreStartTimeout = DateTime.MinValue;

        public event Action<JsonObject> OnStateUpdate;
        public event Action<string, string> OnError;
        public event Action OnDisconnected;
        public event Action OnShutdown;

        public bool IsConnected => _connected;

        public async Task ConnectAsync()
        {
            _cts = new CancellationTokenSource();
            int delay = 100;

            while (!_cts.Token.IsCancellationRequested)
            {
                try
                {
                    _pipe = new NamedPipeClientStream(".", PipeName, PipeDirection.InOut, PipeOptions.Asynchronous);
                    await _pipe.ConnectAsync(1000, _cts.Token);

                    _reader = new StreamReader(_pipe);
                    _writer = new StreamWriter(_pipe) { AutoFlush = true };

                    // Send hello
                    var hello = new JsonObject
                    {
                        ["type"] = "hello",
                        ["version"] = ProtocolVersion,
                        ["name"] = "loupedeck",
                        ["commands"] = new JsonArray("axis", "button", "reset")
                    };
                    await _writer.WriteLineAsync(hello.ToJsonString());

                    // Read welcome
                    var welcomeLine = await _reader.ReadLineAsync();
                    if (welcomeLine == null) throw new IOException("No welcome received");

                    var welcome = JsonNode.Parse(welcomeLine)?.AsObject();
                    if (welcome?["type"]?.GetValue<string>() != "welcome")
                        throw new IOException("Expected welcome message");

                    var apiStatusNode = welcome["apiStatus"]?.AsObject();
                    if (apiStatusNode != null)
                    {
                        foreach (var kvp in apiStatusNode)
                        {
                            var status = kvp.Value?.GetValue<string>();
                            if (status == "deprecated")
                            {
                                System.Diagnostics.Trace.WriteLine($"[Apricadabra] Warning: command '{kvp.Key}' is deprecated");
                            }
                            else if (status == "undefined")
                            {
                                System.Diagnostics.Trace.WriteLine($"[Apricadabra] Error: command '{kvp.Key}' is undefined — core may need upgrade");
                            }
                        }
                    }
                    var coreVersion = welcome["coreVersion"]?.GetValue<string>();
                    if (coreVersion != null)
                    {
                        System.Diagnostics.Trace.WriteLine($"[Apricadabra] Connected to core v{coreVersion}");
                    }

                    _connected = true;
                    delay = 100;

                    // UDP sender for commands
                    _udpSender = new UdpClient();
                    _udpSender.Connect(IPAddress.Loopback, UdpCommandPort);

                    // Start pipe read loop (heartbeat only)
                    _pipeReadTask = Task.Run(() => PipeReadLoopAsync(_cts.Token));

                    // Start UDP listener for state broadcasts
                    _udpListenTask = Task.Run(() => UdpListenLoopAsync(_cts.Token));

                    // Dispatch initial state from welcome
                    OnStateUpdate?.Invoke(welcome);
                    return;
                }
                catch (Exception) when (!_cts.Token.IsCancellationRequested)
                {
                    TryLaunchCore();
                    await Task.Delay(delay, _cts.Token);
                    delay = Math.Min(delay * 2, 5000);
                }
            }
        }

        public async Task SendAsync(JsonObject message)
        {
            if (!_connected || _udpSender == null) return;
            try
            {
                var bytes = Encoding.UTF8.GetBytes(message.ToJsonString());
                await _udpSender.SendAsync(bytes, bytes.Length);
            }
            catch (Exception ex)
            {
                System.Diagnostics.Trace.WriteLine($"[Apricadabra] SendAsync failed: {ex.Message}");
            }
        }

        private async Task PipeReadLoopAsync(CancellationToken ct)
        {
            try
            {
                while (!ct.IsCancellationRequested)
                {
                    var line = await _reader.ReadLineAsync();
                    if (line == null) break;

                    var msg = JsonNode.Parse(line)?.AsObject();
                    if (msg == null) continue;

                    var msgType = msg["type"]?.GetValue<string>();
                    switch (msgType)
                    {
                        case "heartbeat":
                            // Heartbeat ack goes over pipe, not UDP
                            try { await _writer.WriteLineAsync(new JsonObject { ["type"] = "heartbeat_ack" }.ToJsonString()); }
                            catch (Exception ex) { System.Diagnostics.Trace.WriteLine($"[Apricadabra] Heartbeat ack failed: {ex.Message}"); }
                            break;
                        case "error":
                            OnError?.Invoke(
                                msg["code"]?.GetValue<string>() ?? "unknown",
                                msg["message"]?.GetValue<string>() ?? "Unknown error"
                            );
                            break;
                        case "core_restarting":
                            var timeout = msg["coreStartTimeout"]?.GetValue<int>() ?? 15000;
                            System.Diagnostics.Trace.WriteLine($"[Apricadabra] Core restarting, suppressing auto-launch for {timeout}ms");
                            _coreStartTimeout = DateTime.UtcNow.AddMilliseconds(timeout);
                            break;
                        case "shutdown":
                            OnShutdown?.Invoke();
                            return;
                    }
                }
            }
            catch (OperationCanceledException) { }
            catch (Exception ex)
            {
                System.Diagnostics.Trace.WriteLine($"[Apricadabra] Pipe read error: {ex.Message}");
            }

            HandleDisconnect();
        }

        private async Task UdpListenLoopAsync(CancellationToken ct)
        {
            try
            {
                using var udp = new UdpClient(UdpBroadcastPort);
                udp.Client.ReceiveTimeout = 5000;

                while (!ct.IsCancellationRequested)
                {
                    try
                    {
                        var result = await udp.ReceiveAsync(ct);
                        var json = Encoding.UTF8.GetString(result.Buffer);
                        var msg = JsonNode.Parse(json)?.AsObject();
                        if (msg != null && msg["type"]?.GetValue<string>() == "state")
                        {
                            OnStateUpdate?.Invoke(msg);
                        }
                    }
                    catch (OperationCanceledException) { break; }
                    catch (Exception ex)
                    {
                        System.Diagnostics.Trace.WriteLine($"[Apricadabra] UDP receive error: {ex.Message}");
                    }
                }
            }
            catch (OperationCanceledException) { }
            catch (Exception ex)
            {
                System.Diagnostics.Trace.WriteLine($"[Apricadabra] UDP listener error: {ex.Message}");
            }
        }

        private void HandleDisconnect()
        {
            _connected = false;
            OnDisconnected?.Invoke();
            _ = Task.Run(async () =>
            {
                await Task.Delay(1000);
                await ConnectAsync();
            });
        }

        private void TryLaunchCore()
        {
            if (DateTime.UtcNow < _coreStartTimeout)
            {
                return; // Suppress auto-launch during core restart
            }

            try
            {
                var appData = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
                var corePath = Path.Combine(appData, "Apricadabra", CoreExeName);

                if (!File.Exists(corePath))
                {
                    corePath = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, CoreExeName);
                }

                if (File.Exists(corePath))
                {
                    Process.Start(new ProcessStartInfo
                    {
                        FileName = corePath,
                        UseShellExecute = false,
                        CreateNoWindow = true
                    });
                }
            }
            catch (Exception ex)
            {
                System.Diagnostics.Trace.WriteLine($"[Apricadabra] Failed to launch core: {ex.Message}");
            }
        }

        public void Dispose()
        {
            _cts?.Cancel();
            _udpSender?.Dispose();
            _pipe?.Dispose();
            _reader?.Dispose();
            _writer?.Dispose();
        }
    }
}

File: streamdeck-plugin/src/core-connection.ts
import * as net from "net";
import * as dgram from "dgram";
import { spawn } from "child_process";
import { join } from "path";
import { existsSync } from "fs";

const PIPE_NAME = "\\\\.\\pipe\\apricadabra";
const UDP_COMMAND_PORT = 19871;
const UDP_BROADCAST_PORT = 19873; // Unique port for Stream Deck
const PROTOCOL_VERSION = 2;
const CORE_EXE_NAME = "apricadabra-core.exe";

type StateCallback = (axes: Record<string, number>, buttons: Record<string, boolean>) => void;
type StatusCallback = (status: string) => void;

export class CoreConnection {
    private pipe: net.Socket | null = null;
    private udpSender: dgram.Socket | null = null;
    private udpListener: dgram.Socket | null = null;
    private connected = false;
    private reconnecting = false;
    private coreStartTimeoutUntil = 0;
    private buffer = "";

    public onStateUpdate: StateCallback | null = null;
    public onStatusChange: StatusCallback | null = null;

    async connect(): Promise<void> {
        let delay = 100;
        while (true) {
            try {
                await this.tryConnect();
                return;
            } catch {
                this.tryLaunchCore();
                await this.sleep(delay);
                delay = Math.min(delay * 2, 5000);
            }
        }
    }

    private tryConnect(): Promise<void> {
        return new Promise((resolve, reject) => {
            const pipe = net.createConnection(PIPE_NAME);
            let resolved = false;

            pipe.on("connect", () => {
                this.pipe = pipe;
                const hello = JSON.stringify({
                    type: "hello",
                    version: PROTOCOL_VERSION,
                    name: "streamdeck",
                    broadcastPort: UDP_BROADCAST_PORT,
                    commands: ["axis", "button", "reset"],
                });
                pipe.write(hello + "\n");
            });

            pipe.on("data", (data) => {
                this.buffer += data.toString();
                const lines = this.buffer.split("\n");
                this.buffer = lines.pop() || "";

                for (const line of lines) {
                    if (!line.trim()) continue;
                    let msg;
                    try {
                        msg = JSON.parse(line);
                    } catch {
                        continue;
                    }

                    if (!resolved && msg.type === "welcome") {
                        resolved = true;
                        this.connected = true;
                        this.setupUdp();
                        this.onStatusChange?.("Connected");

                        // Parse v2 fields
                        if (msg.apiStatus) {
                            for (const [cmd, status] of Object.entries(msg.apiStatus)) {
                                if (status === "deprecated") {
                                    console.warn(`[Apricadabra] Command '${cmd}' is deprecated`);
                                } else if (status === "undefined") {
                                    console.error(`[Apricadabra] Command '${cmd}' is undefined — core may need upgrade`);
                                    // TODO: Could trigger core_upgrade flow here
                                }
                            }
                        }
                        if (msg.coreVersion) {
                            console.log(`[Apricadabra] Connected to core v${msg.coreVersion}`);
                        }

                        if (msg.axes || msg.buttons) {
                            this.onStateUpdate?.(msg.axes || {}, msg.buttons || {});
                        }
                        resolve();
                    } else if (msg.type === "heartbeat") {
                        pipe.write(JSON.stringify({ type: "heartbeat_ack" }) + "\n");
                    } else if (msg.type === "error") {
                        this.onStatusChange?.(`Error: ${msg.message}`);
                    } else if (msg.type === "shutdown") {
                        this.onStatusChange?.("Core shutting down");
                        this.disconnect();
                    } else if (msg.type === "core_restarting") {
                        const timeout = msg.coreStartTimeout || 15000;
                        console.log(`[Apricadabra] Core restarting, suppressing auto-launch for ${timeout}ms`);
                        this.coreStartTimeoutUntil = Date.now() + timeout;
                        this.onStatusChange?.("Core restarting...");
                    }
                }
            });

            pipe.on("error", (err) => {
                if (!resolved) reject(err);
                else this.handleDisconnect();
            });

            pipe.on("close", () => {
                if (!resolved) reject(new Error("Pipe closed"));
                else this.handleDisconnect();
            });

            setTimeout(() => {
                if (!resolved) {
                    pipe.destroy();
                    reject(new Error("Timeout"));
                }
            }, 2000);
        });
    }

    private setupUdp(): void {
        this.udpSender = dgram.createSocket("udp4");

        this.udpListener = dgram.createSocket({ type: "udp4", reuseAddr: true });
        this.udpListener.bind(UDP_BROADCAST_PORT, "127.0.0.1");
        this.udpListener.on("message", (data) => {
            try {
                const msg = JSON.parse(data.toString());
                if (msg.type === "state") {
                    this.onStateUpdate?.(msg.axes || {}, msg.buttons || {});
                }
            } catch {}
        });
    }

    send(message: Record<string, unknown>): void {
        if (!this.connected || !this.udpSender) return;
        const buf = Buffer.from(JSON.stringify(message));
        this.udpSender.send(buf, UDP_COMMAND_PORT, "127.0.0.1");
    }

    private handleDisconnect(): void {
        if (this.reconnecting) return;
        this.reconnecting = true;
        this.connected = false;
        this.onStatusChange?.("Disconnected");
        this.cleanup();
        setTimeout(() => {
            this.reconnecting = false;
            this.connect();
        }, 1000);
    }

    private disconnect(): void {
        this.connected = false;
        this.cleanup();
    }

    private cleanup(): void {
        this.pipe?.destroy();
        this.pipe = null;
        this.udpSender?.close();
        this.udpSender = null;
        this.udpListener?.close();
        this.udpListener = null;
    }

    private tryLaunchCore(): void {
        if (Date.now() < this.coreStartTimeoutUntil) {
            return; // Suppress auto-launch during core restart
        }
        const appData = process.env.APPDATA || "";
        const candidates = [
            join(appData, "Apricadabra", CORE_EXE_NAME),
            join(__dirname, CORE_EXE_NAME),
            join(__dirname, "..", CORE_EXE_NAME),
        ];
        for (const corePath of candidates) {
            if (existsSync(corePath)) {
                spawn(corePath, [], { detached: true, stdio: "ignore" }).unref();
                return;
            }
        }
    }

    private sleep(ms: number): Promise<void> {
        return new Promise((r) => setTimeout(r, ms));
    }
}
